### ISSUE TITLE
RES ActiveRecord uses unnecessary join 

### ISSUE THREAD

**Issue by leifg** [RailsEventStore/rails_event_store#1517]
When reading from a stream with default filters, there is no need to join the events table on the stream table because all available attributes are available on `event_store_events_in_streams`.

However the reader [always joins the events table](https://github.com/RailsEventStore/rails_event_store/blob/master/ruby_event_store-active_record/lib/ruby_event_store/active_record/event_repository_reader.rb#L101)

With a large events table and reading a lot of events from a stream (in our case 30,000+) this significantly slows down reading from a stream.

Unfortunately you can't just add a guard around the order method because the filter on the events table is needed further down

```ruby
stream = ordered(stream.joins(:event), spec) if spec.time_sort_by
stream = stream.order(id: order(spec))
stream = stream.limit(spec.limit) if spec.limit?
stream = stream.where(start_condition(spec)) if spec.start
stream = stream.where(stop_condition(spec)) if spec.stop
stream = stream.where(older_than_condition(spec)) if spec.older_than # <== fails here
stream = stream.where(older_than_or_equal_condition(spec)) if spec.older_than_or_equal
stream = stream.where(newer_than_condition(spec)) if spec.newer_than
stream = stream.where(newer_than_or_equal_condition(spec)) if spec.newer_than_or_equal
```

So probably a generic guard that predicts all filters on the event table would make the most sense:

```ruby
stream = stream.joins(:event) if needs_join?(spec)

def needs_join?(spec)
  spec.with_types? || spec.time_sort_by || spec.older_than || spec.older_than_or_equal || spec.newer_than || spec.newer_than_or_equal 
end
```

Or is there a more elegant solution that this?

**Comment by fidel** [comment]
Hey @leifg! 

Could you share a bit more details about your use case? Reading 30k+ events sounds a lot. 

Providing some setup information like DB and exact SQL which is generated would be also helpful. 

`event_store_events_in_streams` table keeps only event's ids, which is not enough to return read events, join to `event_store_events` is required anyways at some point.

**Comment by leifg** [comment]
> Could you share a bit more details about your use case? Reading 30k+ events sounds a lot.

Sorry I should have been more clear. This is not a use but a situation in our system. We have one stream with 30k events. This happened by accident over a couple of months. The result is: the whole system is slowing down whenever we touch this stream. There are multiple ways to optimize this in we are certainly gonna do more than one thing to get rid of this issue (one thing is probably going to get rid of the stream all together and rebuild the data).

That being said, there is one optimization that can happen within the `ruby_event_store-active_record`

> event_store_events_in_streams table keeps only event's ids, which is not enough to return read events, join to event_store_events is required anyways at some point.

That is true if you want to (efficiently) filter on attributes in the events table such as event type, timestamp etc...

However if you want to just load all events in a stream in order all you need to do is do 2 seperate queries. One to load the rows from the stream table, and then one query to query all events. This is happening via the [`.preload(:event)`](https://github.com/RailsEventStore/rails_event_store/blob/master/ruby_event_store-active_record/lib/ruby_event_store/active_record/event_repository_reader.rb#L98) in the first row]:

```ruby
irb(main):050:0> @stream_klass.where(stream: "my_stream").preload(:event)

# SELECT "event_store_events_in_streams".* FROM "event_store_events_in_streams" WHERE "event_store_events_in_streams"."stream" = $1
# SELECT "event_store_events".* FROM "event_store_events" WHERE "event_store_events"."event_id" IN ($1, $2)
```

So for certain query the join is unnecessary.

**Comment by fidel** [comment]
Thanks for the broader explanation @leifg, it helped us understand what's the root of the problem.

The fix of issue you have exposed is on master branch already, official release should be expected soon.

Cheers!

**Comment by lukaszreszke** [comment]
Hi @leifg,

We have just released version 2.8.0 containing the fix for the issue that you have reported. 

We'll appreciate any feedback.

**Comment by leifg** [comment]
That was quick. Thank you so much!
